fdbclient/NativeAPI.actor.cpp

```diff
@@ -3176,6 +3176,46 @@ Future<Void> Transaction::onError( Error const& e ) {
 
 	return e;
 }
+ACTOR Future<StorageMetrics> getStorageMetricsLargeKeyRange(Database cx, KeyRangeRef keys);
+
+ACTOR Future<StorageMetrics> doGetStorageMetrics(Database cx, KeyRangeRef keys, Reference<LocationInfo> locationInfo) {
+	loop {
+		try {
+			WaitMetricsRequest req(keys, StorageMetrics(), StorageMetrics());
+			req.min.bytes = 0;
+			req.max.bytes = -1;
+			StorageMetrics m = wait(
+			    loadBalance(locationInfo, &StorageServerInterface::waitMetrics, req, TaskPriority::DataDistribution));
+			return m;
+		} catch (Error& e) {
+			if (e.code() != error_code_wrong_shard_server && e.code() != error_code_all_alternatives_failed) {
+				TraceEvent(SevError, "WaitStorageMetricsError").error(e);
+				throw;
+			}
+			cx->invalidateCache(keys);
+			StorageMetrics m = wait(getStorageMetricsLargeKeyRange(cx, keys));
```

Comment:
I notice another implementation of this loop has a delay here. Are there failure cases where this is going to spin too fast?

Reply:
I don't think so. This `catch` is mainly to deal with the case where the target shard got split during collecting the byte sample. I think that'll be pretty rare.

Reply:
Ok, I see. It looks like the `all_alternatives_failed` case is introducing its own delay in LoadBalance in the case that it gets thrown, and it does seem like it would be rare to see very many repeats in `wrong_shard_server`. Because I'm not super confident that I know all the failure cases here, though, I'll also ping @etschannen to make sure there's nothing I'm missing.

Reply:
This request is not versioned, so if a storage server is behind it might not know it has been assigned a shard range that a proxy thinks it has, so you could get into a loop, it is worth adding the delay just in case

Reply:
Delay added.